src/diff.ts

```diff
@@ -0,0 +1,103 @@
+import { PropertyChangeRecord } from './interfaces';
+
+export const enum DiffType {
+	CUSTOM = 1,
+	IGNORE,
+	REFERENCE,
+	SHALLOW,
+	AUTO
+}
+
+function isObjectOrArray(value: any): boolean {
+	return Object.prototype.toString.call(value) === '[object Object]' || Array.isArray(value);
+}
+
+function ignore(previousProperty: any, newProperty: any): PropertyChangeRecord {
+	return {
+		changed: false,
+		value: newProperty
+	};
+}
+
+function custom(previousProperty: any, newProperty: any, meta: any): PropertyChangeRecord {
```

Comment:
the `meta` is meant to be a generic value bag. we only have one usage of it at the moment which does have `diffFunction` and `scope`, but thats extremely coupled to the existing functionality of `diffProperty`. I think we'll revisit the decorator functions in https://github.com/dojo/widget-core/issues/402, so I think we should leave as is for now (given its internal anyway).